b2/src/main/java/org/jclouds/b2/blobstore/B2BlobStore.java

```diff
@@ -0,0 +1,480 @@
+      Set<String> commonPrefixes = Sets.newHashSet();
+      for (B2ObjectList.Entry entry : list.files()) {
+         // B2 does not support server-side filtering via prefix and delimiter so we emulate it on the client.
+         // TODO: cooperate with maxResults
+         if (options.getPrefix() != null && !entry.fileName().startsWith(options.getPrefix())) {
+            continue;
+         }
+
+         if (delimiter != null) {
+            String fileName = entry.fileName();
+            int index = entry.fileName().indexOf(delimiter, Strings.nullToEmpty(options.getPrefix()).length());
+            if (index != -1) {
+               String prefix = entry.fileName().substring(0, index + 1);
+               if (!commonPrefixes.contains(prefix)) {
+                  commonPrefixes.add(prefix);
+                  builder.add(new StorageMetadataImpl(StorageType.RELATIVE_PATH, null, prefix, null, null, null, null, null, ImmutableMap.<String, String>of(), null));
+               }
+               continue;
+            }
+         }
+
+         if (options.isDetailed()) {
+            BlobMetadata metadata = blobMetadata(container, entry.fileName());
+            if (metadata != null) {
+               builder.add(metadata);
+            }
+         } else {
+            Map<String, String> userMetadata = ImmutableMap.of();
+            ContentMetadata metadata = ContentMetadataBuilder.create()
+                  .contentLength(entry.size())
+                  .build();
+            builder.add(new BlobMetadataImpl(null, entry.fileName(), null, null, null, null, entry.uploadTimestamp(), userMetadata, null, container, metadata, entry.size()));
+         }
+      }
+      return new PageSetImpl<StorageMetadata>(builder.build(), list.nextFileName());
+   }
+
+   @Override
+   public boolean blobExists(String container, String name) {
+      return blobMetadata(container, name) != null;
+   }
+
+   @Override
+   public String putBlob(String container, Blob blob) {
+      return putBlob(container, blob, PutOptions.NONE);
+   }
+
+   @Override
+   public String putBlob(String container, Blob blob, PutOptions options) {
+      Preconditions.checkArgument(blob.getPayload().isRepeatable(), "B2 requires repeatable payload to calculate SHA1");
+
+      if (options.getBlobAccess() != BlobAccess.PRIVATE) {
+         throw new UnsupportedOperationException("B2 only supports private access blobs");
+      }
+
+      if (options.isMultipart()) {
+         return putMultipartBlob(container, blob, options);
+      } else {
+         String contentSha1;
+         try {
+            contentSha1 = ByteStreams2.hashAndClose(blob.getPayload().openStream(), Hashing.sha1()).toString();
+         } catch (IOException ioe) {
+            throw Throwables.propagate(ioe);
+         }
+
+         String name = blob.getMetadata().getName();
+
+         // B2 versions all files so we store the original fileId to delete it after the upload succeeds
```

Comment:
I see. Should we be doing this, though? I am not sure, from a user perspective, what the expected behavior would be.